doc/source/notebooks/manual/tailor/mixture_density_network.ipynb

```diff
@@ -4,15 +4,745 @@
+    "$$\n",
+    "Each of the parameters  $\\pi_{m}(x), \\mu_{m}(x), \\sigma_{m}(x)$ of the distribution will be determined by the neural network, as a function of the input $x$.\n",
+    "\n",
+    "We train the MDN's neural net by optimizing the model's likelihood\n",
+    "$$\n",
+    " \\mathcal{L} \\triangleq \\operatorname{argmax}_{\\Theta} \\prod_{n=1}^N p(Y = y_n\\,|\\,X = x_n),\n",
+    "$$\n",
+    "where $\\Theta$ collects the neural nets weights and biases and $\\{x_n, y_n\\}_{n=1}^N$ are our training dataset."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## GPflow MDN implementation\n",
+    "\n",
+    "GPflow doesn't reinvent the wheel: most of what you will see is just plain Python/TensorFlow code. We choose to use GPflow, however, because it provides us with some functionality to easily define a model. Once we have a GPflow model, we can specify its objective function, parameters, dataset. This extra layer of abstraction makes interacting with the model, such as optimizing or performing inference, much easier. \n",
+    "\n",
+    "We begin by importing the required packages from TensorFlow and GPflow"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 3,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import tensorflow as tf\n",
+    "from scipy.stats import norm"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 4,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import gpflow\n",
+    "from gpflow.models.model import Model\n",
+    "from gpflow.params.dataholders import Minibatch, DataHolder\n",
+    "from gpflow.params import Parameter, ParamList\n",
+    "from gpflow.training import AdamOptimizer, ScipyOptimizer\n",
+    "from gpflow.decors import params_as_tensors, autoflow\n",
+    "\n",
+    "float_type = gpflow.settings.float_type"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We continue by creating a `MDN` class which inherits from GPflow's `model` class. We now need to do the following things:\n",
+    "1. Store the feature and target matrices (X, Y) as `DataHolders`.\n",
+    "2. Define our model's parameters using GPflow's `Parameter` and `ParamList` objects.\n",
+    "3. Define the objective function. In the `_build_likelood` function we need to specify our model's objective function. When we optimize the model the negative of this function will be minimized. "
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class MDN(Model):\n",
+    "    \n",
+    "    def __init__(self, X, Y, inner_dims=[10, 10,], activation=tf.nn.tanh, num_mixtures=5):\n",
+    "        Model.__init__(self)\n",
+    "        \n",
+    "        self.Din = X.shape[1]\n",
+    "        self.dims = [self.Din, ] + list(inner_dims) + [3 * num_mixtures]\n",
```

Comment:
Can you clarify what happens in this line? A comment or an explanation in the cell would be good.